core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala

```diff
@@ -287,11 +258,10 @@ private[spark] object TaskMetrics extends Logging {
   def fromAccumulatorUpdates(accumUpdates: Seq[AccumulableInfo]): TaskMetrics = {
     val definedAccumUpdates = accumUpdates.filter(_.update.isDefined)
     val metrics = new ListenerTaskMetrics(definedAccumUpdates)
-    // We don't register this [[ListenerTaskMetrics]] for cleanup, and this is only used to post
-    // event, we should un-register all accumulators immediately.
-    metrics.internalAccums.foreach(acc => Accumulators.remove(acc.id))
-    definedAccumUpdates.filter(_.internal).foreach { accum =>
-      metrics.internalAccums.find(_.name == accum.name).foreach(_.setValueAny(accum.update.get))
+    definedAccumUpdates.filter(_.internal).foreach { accInfo =>
+      metrics.internalAccums.find(_.name == accInfo.name).foreach { acc =>
+        acc.asInstanceOf[Accumulator[Any, Any]].add(accInfo.update.get)
```

Comment:
Just have a reset and "add"? 

I'd argue it doesn't make sense to call setValue, since "set" action is not algebraic (i.e. you cannot compose/merge set operations).